### PR TITLE
Improving speed over overal type-checking by factor 2

### PR DESCRIPTION
**Pull request by DavyLandman**
I tested this, and it reduced the runtime for typechecking the parsergenerator from 800s to 350s.

I think there might be some more performance gains to make by using a smarter datastructure to detect common types (a rel comes to mind) but I leave that for you to think about.

### CODE REVIEW

**Comment by jurgenvinju** [comment]
This is exciting.

File: src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc [hunk 1]
```diff
@@ -353,7 +353,8 @@ void checkOverloading(map[str,Tree] namedTrees, Solver s){
     }
     try {
     consDef = range(consNameDef);
-    for(d1 <- consDef, d2 <- consDef, d1.defined != d2.defined, t1 := s.getType(d1), t2 := s.getType(d2), t1.adt == t2.adt,
+    consTypes = (d : s.getType(d) | d <- consDef);
+    for(d1 <- consDef, d2 <- consDef, d1.defined != d2.defined, t1 := consTypes[d1], t2 := consTypes[d1], t1.adt == t2.adt,
```
Comment (jurgenvinju) [inline src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc:357]: There's a typo in `t2 := consTypes[d1]`, t2 should be defined by the constypes of d2 ?

Comment (jurgenvinju) [reply src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc:357]: Might have to run the measurements again.

Comment (jurgenvinju) [reply src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc:357]: Also if this loop is hot then introducing types for the variable bindings will really speed things up. It removes the need to traverse the scope stack for finding bound variables with the same name.

**Review by jurgenvinju — CHANGES_REQUESTED** [review]

**Comment by PaulKlint** [comment]
Good finding driven by profiling!
Disclosure: there may be dozens of such loops. Designing the logic and optimizing it at the same time does not fit in my head most of the time.

File: src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc (hunk 1 above)
Comment (DavyLandman) [reply src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc:357]: ah sorry, I did it in the browser, my bad, I ran the test with the correct edit.

**Comment by DavyLandman** [comment]
sure, let's only do this if it's slow ;) not prematurely.

**Comment by DavyLandman** [comment]
Note it's still one of the slower parts, that specific loop. I think if we store it into a rel, we can drop the equality checks and use the multimap properties for it (instead cross product loop).

**Comment by jurgenvinju** [comment]
If this increases speed by 2 then the loop must be very hot. Let's add types to the pattern variables now that we are here anyway and then merge the branch. Next stop: getting the relation type to work for us like Davy said.

**Comment by jurgenvinju** [comment]
> sure, let's only do this if it's slow ;) not prematurely.

if you are talking about the variable types (it's not visible from the current github overview of this PR), then I would _not_ say this is premature. We know the loop is hot. We know that untyped bindings will lead to a factor of additional interpretive overhead _for each match_. The factor depends on the depth of the scope stack and the number of imported modules. 

So let's just add the variables and benefit from even faster code. Not premature.

**Comment by DavyLandman** [comment]
My response was about premature optimization, Paul warned there are many like this, I just meant: we just rewrite/improve the slow ones.

This morning in the shower I realized how to make it a lot quicker still. Will see when I can write that out.

**Comment by DavyLandman** [comment]
I added the latest version, please review again.

It brings it down to 293s (and finally bumps down `isOverloadedFunction` from it's first position
(now a differen overload of that function is the most common one).

```
FRAMES PROFILE: 6104 data points, 293194 ticks, tick = 1 milliSecs
                                Scope   Ticks        %  Source
                            newSolver   51041    17.4%  |lib://typepal/src/analysis/typepal/Solver.rsc|(454,69652,<25,0>,<1782,1>)
                        newScopeGraph   48902    16.7%  |lib://typepal/src/analysis/typepal/ConfigurableScopeGraph.rsc|(6703,13747,<178,0>,<499,1>)
                 isOverloadedFunction   36669    12.5%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc|(9112,364,<224,0>,<232,1>)
                           addGrammar   21437     7.3%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/ADTandGrammar.rsc|(4904,5570,<109,0>,<228,1>)
                           saveModule   14174     4.8%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/Import.rsc|(12479,6041,<269,0>,<382,1>)
             rascalIsAcceptableSimple   10487     3.6%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc|(1916,1276,<64,0>,<87,1>)
                         newCollector    7952     2.7%  |lib://typepal/src/analysis/typepal/Collector.rsc|(8291,35339,<188,0>,<1014,1>)
                              addADTs    7157     2.4%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/ADTandGrammar.rsc|(562,352,<27,0>,<33,1>)
                   rascalReportUnused    6247     2.1%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc|(9478,2876,<234,0>,<290,1>)
                     checkOverloading    3214     1.1%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc|(12791,3964,<300,0>,<368,1>)
              getImportAndExtendGraph    3128     1.1%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/Import.rsc|(6396,4045,<142,0>,<220,1>)
                           isSameFile    3108     1.1%  |lib://rascal/Location.rsc|(885,230,<32,0>,<40,5>)

ASTS PROFILE: 6104 data points, 293194 ticks, tick = 1 milliSecs
                            Scope   Ticks        %  Source
             isOverloadedFunction   19192     6.5%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc|(9346,1,<228,66>,<228,67>)
                  lookupPathsWide   14516     5.0%  |lib://typepal/src/analysis/typepal/ConfigurableScopeGraph.rsc|(14683,5,<362,49>,<362,54>)
                             _run   10796     3.7%  |lib://typepal/src/analysis/typepal/Solver.rsc|(66609,7,<1707,61>,<1707,68>)
             isOverloadedFunction    9248     3.2%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc|(9360,5,<228,80>,<228,85>)
                       addGrammar    8203     2.8%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/ADTandGrammar.rsc|(8136,7,<182,31>,<182,38>)
                       saveModule    7790     2.7%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/Import.rsc|(17092,2,<353,19>,<353,21>)
                       addGrammar    4879     1.7%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/ADTandGrammar.rsc|(5719,1,<123,66>,<123,67>)
                          addADTs    4052     1.4%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/ADTandGrammar.rsc|(789,1,<31,17>,<31,18>)
                         bindWide    3819     1.3%  |lib://typepal/src/analysis/typepal/ConfigurableScopeGraph.rsc|(13720,2,<340,19>,<340,21>)
                          isEmpty    3617     1.2%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/Import.rsc|(17122,1,<353,49>,<353,50>)
         rascalIsAcceptableSimple    3528     1.2%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc|(3174,15,<86,12>,<86,27>)
                       addGrammar    3471     1.2%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/ADTandGrammar.rsc|(5318,1,<116,64>,<116,65>)
                       addGrammar    3234     1.1%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/ADTandGrammar.rsc|(8213,1,<182,108>,<182,109>)
         rascalIsAcceptableSimple    3085     1.1%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/RascalConfig.rsc|(3066,17,<83,19>,<83,36>)
                          addADTs    3059     1.0%  |file:///export/scratch1/landman/rascal-stuff/rascal-core/src/org/rascalmpl/core/library/lang/rascalcore/check/ADTandGrammar.rsc|(857,1,<31,85>,<31,86>)
list[ModuleMessages]: [
```

**Comment by jurgenvinju** [comment]
Reviewed. This is good progress and pivotal. Let's work on optimizing the next bottlenecks in parallel when we have some time. This can be merged and released for the other projects to profit from.

**Pull request merged**